Abstracted robot collision checking into class for snap and vectorfield

cdellin opened the pull request:
In the push for baked checking in PrPy, this is a first pass which creates a two robot collision checkers:
- `SimpleRobotCollisionChecker`
- `BakedRobotCollisionChecker`
Each of these can be constructed for a robot, and expose a simple `CheckCollision` method, which raises a PrPy exception upon collision.

This PR also refactors the `SnapPlanner` and `VectorFieldPlanner` planners to make use of this abstraction.  The planners default to `SimpleRobotCollisionChecker`, but a different type can be passed in when they are instantiated, and this type will be used for its collision checking.

Note that the `BakedRobotCollisionChecker` is experimental, and only works with the `baked_checkers` branch of `or_fcl` at the moment.  Currently, the underlying checker does not expose which links provoked the collision, so the collision report is not helpful, and a vanilla `CollisionPlanningError` is currently thrown.

I'll wait to implement this for `TSRPlanner` until its PR (#322) is merged.

mkoval commented on src/prpy/collision.py:
```diff
@@ -0,0 +1,65 @@
+from prpy.planning.exceptions import SelfCollisionPlanningError
+
+
+class SimpleRobotCollisionChecker:
```
Missing docstring.

cdellin replied:
Added docstrings.

mkoval commented on src/prpy/collision.py:
```diff
@@ -0,0 +1,65 @@
+    def __init__(self, robot):
+        self.robot = robot
+        self.env = robot.GetEnv()
+        self.checker = self.env.GetCollisionChecker()
```
Can you check the type of `GetCollisionChecker` before trying to use it?

cdellin replied:
I don't check the type, but I do now check for the case where no checker exists, and I raise a more helpful exception when the checker doesn't support baked checking.

mkoval commented on src/prpy/collision.py:
```diff
@@ -0,0 +1,65 @@
+        self.checker.SendCommand('BakeBegin')
+        self.env.CheckCollision(self.robot)
+        self.robot.CheckSelfCollision()
+        self.baked = openravepy.RaveCreateKinBody(self.env, kb_type)
```
Check for an error here.

cdellin replied:
Once we've confirmed that the checker supports baking, this is assumed to work.  If not, it already throws an exception (kinbody type not known!) which is correct.

mkoval commented on src/prpy/collision.py:
```diff
@@ -0,0 +1,65 @@
+
+    def CheckCollision(self):
+        report = openravepy.CollisionReport()
+        if self.checker.CheckSelfCollision(self.baked, report):
```
Please add a comment explaining that this performs the baked check.

cdellin replied:
Comment added!

mkoval commented on src/prpy/collision.py:
```diff
@@ -0,0 +1,65 @@
+        kb_type = self.checker.SendCommand('BakeGetType')
+        self.checker.SendCommand('BakeBegin')
+        self.env.CheckCollision(self.robot)
+        self.robot.CheckSelfCollision()
```
Please add a comment explaining that this bakes the two collision checks.

cdellin replied:
Comment added.

mkoval commented on src/prpy/planning/snap.py:
```diff
@@ -36,6 +36,7 @@
     ClonedPlanningMethod,
     Tags
 )
+from prpy.collision import SimpleRobotCollisionChecker
```
Prefer: `from ..collision import SimpleRobotCollisionChecker`

cdellin replied:
Fixed (against my better judgment `(-:`).

mkoval commented on src/prpy/planning/vectorfield.py:
```diff
@@ -35,6 +35,7 @@
 import openravepy
 from .. import util
 from base import BasePlanner, PlanningError, ClonedPlanningMethod, Tags
+from prpy.collision import SimpleRobotCollisionChecker
```
Prefer: `from ..collision import SimpleRobotCollisionChecker`

mkoval commented:
Looks good, other than the minor comments I left above. Can you also add this to CHOMP, GreedyIk, and TSRPlanner?

cdellin commented:
I addressed most of the comments, and also added this for `ChompPlanner`.

I'd prefer to wait to add this to `TSRPlanner` until #322 is merged.  Also, the current implementation of `GreedyIKPlanner` does its collision checking in `FindIKSolutions`, and unfortunately there's no way to add baked support to that planner.

cdellin commented:
I like how immediately after my last comment, you went ahead and merged #322.  Fine!  `TSRPlanner` support is now in.  (I refactored the robot collision interface to make it a bit nicer and more consistent with the native OpenRAVE interface.)

mkoval commented on src/prpy/collision.py:
```diff
@@ -0,0 +1,108 @@
+        # The baked check is performed by checking self collision on baked
+        if self.checker.CheckSelfCollision(self.baked, report):
+            return True
+        return False
```
Replace this with:
```python
return self.checker.CheckSelfCollision(self.baked, report)
```
:grin:

cdellin replied:
Haha done!

mkoval commented on src/prpy/planning/tsr.py:
```diff
@@ -136,12 +138,16 @@ def compute_ik_solutions(tsrchain):
+        with CollisionOptionsStateSaver(
+                self.env.GetCollisionChecker(), CollisionOptions.ActiveDOFs):
+            robot_checker = self.robot_collision_checker(robot)
+
```
You should move the `SetActiveDOFs` call above the baking.

cdellin replied:
Good catch! Fixed.

cdellin commented:
Comments addressed!
